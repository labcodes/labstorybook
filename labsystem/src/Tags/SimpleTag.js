import React from "react";
import PropTypes from "prop-types";
import { isEmpty } from "lodash";
import AbstractTag from "./AbstractTag";
import Icon from "../Icon";

export default class SimpleTag extends React.Component {
  static propTypes = {
    text: PropTypes.string.isRequired,
    thumbSrc: PropTypes.string,
    icon: PropTypes.string,
    isOutline: PropTypes.bool,
    skin: PropTypes.string,
    color: PropTypes.string,
    disabled: PropTypes.bool,
  };

  static defaultProps = {
    thumbSrc: "",
    icon: "",
    isOutline: false,
    skin: "pale",
    color: "",
    disabled: false,
  };

  constructor(props) {
    super(props);
    this.checkThumbAndIcon();
  }

  componentDidUpdate() {
    this.checkThumbAndIcon();
  }

  thumb = () => {
    const { thumbSrc } = this.props;
    return thumbSrc ? (
      <img className="lab-tag__thumb" src={thumbSrc} alt="" />
    ) : undefined;
  };

  icon = () => {
    const { icon } = this.props;
    return icon ? (
      <Icon type={icon} color="black75" size="petit" className="left-icon" />
    ) : undefined;
  };

  checkThumbAndIcon() {
    const errorMessage =
      "`SimpleTag` can't be initialized with both `thumb` and `icon` props.";
    const { thumbSrc, icon } = this.props;
    if (!isEmpty(thumbSrc) && !isEmpty(icon)) {
      throw new Error(errorMessage);
    }
  }

  render() {
    const {
      text,
      thumbSrc,
      icon,
      isOutline,
      skin,
      color,
      disabled,
    } = this.props;
    return (
      <AbstractTag
        text={text}
        className={`${icon ? ` lab-tag--has-left-icon` : ""}${
          thumbSrc ? ` lab-tag--has-thumb` : ""
        }`}
        isOutline={isOutline}
        skin={skin}
        color={color}
        disabled={disabled}
        renderPrefix={this.icon() || this.thumb()}
      />
    );
  }
}
