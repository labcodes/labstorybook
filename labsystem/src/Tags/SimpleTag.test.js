import React from "react";
import renderer from "react-test-renderer";
import { mount, shallow } from "enzyme";

import SimpleTag from "./SimpleTag";

describe("SimpleTag", () => {
  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(<SimpleTag text="Test SimpleTag" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected with disabled as true", async () => {
    const renderedComponent = renderer
      .create(<SimpleTag text="Test render disabled SimpleTag" isDisabled />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <SimpleTag text="Test disabled SimpleTag" isDisabled />
    ).html();
    expect(wrapper).toContain("lab-tag--disabled");
  });

  it("renders as expected with outline as true", async () => {
    const renderedComponent = renderer
      .create(<SimpleTag text="Test render outline SimpleTag" isOutline />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <SimpleTag text="Test outline SimpleTag" isOutline />
    ).html();
    expect(wrapper).toContain("lab-tag--outline");
  });

  it("renders as expected with a vivid skin", async () => {
    const renderedComponent = renderer
      .create(<SimpleTag text="Test render vivid SimpleTag" skin="vivid" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <SimpleTag text="Test vivid SimpleTag" skin="vivid" />
    ).html();
    expect(wrapper).toContain("lab-tag--vivid");
  });

  it("renders as expected with a purple color", async () => {
    const renderedComponent = renderer
      .create(<SimpleTag text="Test render purple SimpleTag" color="purple" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <SimpleTag text="Test vivid SimpleTag" skin="vivid" />
    ).html();
    expect(wrapper).toContain("lab-tag--vivid");
  });

  it("renders as expected with a lupe icon", async () => {
    const renderedComponent = renderer
      .create(
        <SimpleTag text="Test render SimpleTag with lupe icon" icon="lupe" />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const wrapper = shallow(
      <SimpleTag text="Test SimpleTag with lupe icon" icon="lupe" />
    ).html();
    expect(wrapper).toContain("lab-icon--lupe");
  });

  const thumbSrcSample =
    "https://avatars3.githubusercontent.com/u/1887591?s=400&u=ba45b6433752099210bf180b4448fb82e015c3a8&v=4";
  it("renders as expected with a thumb", async () => {
    const renderedComponent = renderer
      .create(
        <SimpleTag
          text="Test render SimpleTag with thumb"
          thumbSrc={thumbSrcSample}
        />
      )
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <SimpleTag
        text="Test mount SimpleTag with thumb"
        thumbSrc={thumbSrcSample}
      />
    );
    expect(mountedComponent.find("SimpleTag").prop("thumbSrc")).toEqual(
      thumbSrcSample
    );
  });

  it("does not render if passing both `thumb` and `icon` props", async () => {
    expect(() => {
      shallow(
        <SimpleTag
          text="Test to not render SimpleTag with with thumb and icon"
          icon="lupe"
          thumbSrc={thumbSrcSample}
        />
      );
    }).toThrow(
      "`SimpleTag` can't be initialized with both `thumb` and `icon` props."
    );
  });
});
