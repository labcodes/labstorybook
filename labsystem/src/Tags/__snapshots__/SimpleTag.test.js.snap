// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleTag renders as expected with a lupe icon 1`] = `
<span
  className="lab-tag  lab-tag--has-left-icon lab-tag--pale"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="presentation"
  tabIndex="-1"
>
  <span
    className="lab-icon lab-icon--lupe lab-icon--black-75 lab-icon--petit lab-tag--left-icon"
  />
  Test render SimpleTag with lupe icon
</span>
`;

exports[`SimpleTag renders as expected with a purple color 1`] = `
<span
  className="lab-tag  lab-tag--purple-pale"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="presentation"
  tabIndex="-1"
>
  Test render purple SimpleTag
</span>
`;

exports[`SimpleTag renders as expected with a thumb 1`] = `
<span
  className="lab-tag  lab-tag--has-thumb lab-tag--pale"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="presentation"
  tabIndex="-1"
>
  <img
    alt=""
    className="lab-tag__thumb"
    src="fake-thumb"
  />
  Test render SimpleTag with thumb
</span>
`;

exports[`SimpleTag renders as expected with a vivid skin 1`] = `
<span
  className="lab-tag  lab-tag--vivid"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="presentation"
  tabIndex="-1"
>
  Test render vivid SimpleTag
</span>
`;

exports[`SimpleTag renders as expected with disabled as true 1`] = `
<span
  className="lab-tag  lab-tag--disabled lab-tag--pale"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="presentation"
  tabIndex="-1"
>
  Test render disabled SimpleTag
</span>
`;

exports[`SimpleTag renders as expected with outline as true 1`] = `
<span
  className="lab-tag  lab-tag--outline lab-tag--pale"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="presentation"
  tabIndex="-1"
>
  Test render outline SimpleTag
</span>
`;

exports[`SimpleTag renders with base props 1`] = `
<span
  className="lab-tag  lab-tag--pale"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="presentation"
  tabIndex="-1"
>
  Test render SimpleTag
</span>
`;
